python/lsst/pipe/base/_quantumContext.py

```diff
@@ -39,8 +43,79 @@
 _LOG = getLogger(__name__)
 
 
-class ButlerQuantumContext:
-    """A Butler-like class specialized for a single quantum.
+@dataclass(frozen=True, kw_only=True)
+class ExecutionResources:
+    """A description of the resources available to a running quantum."""
+
+    num_cores: int = 1
+    """The maximum number of cores that the task can use."""
+
+    max_mem: u.Quantity | numbers.Real | None = None
+    """If defined, the amount of memory allocated to the task. If a plain
+    integer is given it is assumed to be the number of bytes and converted
+    to a `~astropy.units.Quantity`.
+    """
+
+    def __post_init__(self) -> None:
+        # Normalize max_mem to bytes.
+        max_mem = self.max_mem
+        if max_mem is not None:
+            if isinstance(max_mem, numbers.Real):
+                max_mem *= u.B
+            else:
+                max_mem = max_mem.to(u.B)
+            object.__setattr__(self, "max_mem", max_mem)
+
+    def __deepcopy__(self, memo: Any) -> ExecutionResources:
+        """Deep copy returns itself because the class is frozen."""
+        return self
+
+    def _reduce_kwargs(self) -> dict[str, Any]:
+        """Return a dict of the keyword arguments that should be used
+        by `__reduce__`.
+
+        This is necessary because the dataclass is defined to be keyword
+        only and we wish the default pickling to only store a plain number
+        for the memory allocation and not a large Quantity.
+
+        Returns
+        -------
+        kwargs : `dict`
+            Keyword arguments to be used when pickling.
+        """
+        kwargs: dict[str, Any] = {"num_cores": self.num_cores}
+        if self.max_mem is not None:
+            # .value is a numpy float. Cast it to a python int since we
+            # do not want fractional bytes.
+            # __post_init__ ensures that this is a Quantity but mypy cannot
+            # work that out.
+            kwargs["max_mem"] = int(self.max_mem.value)  # type: ignore[union-attr]
```

Comment:
Would it be better to declare `self.max_mem` as `u.Quantity | None` and provide a factory method from `max_mem: int`? Then mypy will be happy and you don't need `__post_init__`.

Reply:
Oh, you mean have `ExecutionResources.from_dict(kwargs: dict[str, Any])` or something? (having trouble with the name). I guess I can move the logic from cmdLineFwk here so long as there is a default unit specified for the plain float/int in this constructor. I still want the pickling to store the plain int but I can change the pickle factory to use the new method.

Reply:
MAybe something like `ExecutionResources.from_bytes(num_proc: int, mem_bytes: int)`?

Reply:
If I'm having a factory method I may as well accept strings as well and then it seems odd to have one per type of `max_mem` value. If this factory takes `str|int|Quantity` and converts to Quantity then the ctrl_mpexec code only ever has to call that one API and doesn't care where the max_mem value came from.

Reply:
@andy-slac I've implemented the `__init__` approach as discussed on Slack. If you want to look you can since I'm going to have to run jenkins again.